finality-aleph/src/data_io/chain_info.rs

```diff
@@ -21,82 +19,68 @@ pub trait ChainInfoProvider: Send + Sync + 'static {
+    H: Header,
+    HB: HeaderBackend<H> + 'static,
 {
-    client: Arc<C>,
-    _phantom: PhantomData<B>,
+    client: HB,
+    _phantom: PhantomData<H>,
 }
 
-impl<B, C> SubstrateChainInfoProvider<B, C>
+impl<H, HB> SubstrateChainInfoProvider<H, HB>
 where
-    B: BlockT<Hash = BlockHash>,
-    B::Header: HeaderT<Number = BlockNumber>,
-    C: HeaderBackend<B>,
+    H: Header,
+    HB: HeaderBackend<H>,
 {
-    pub fn new(client: Arc<C>) -> Self {
+    pub fn new(client: HB) -> Self {
         SubstrateChainInfoProvider {
             client,
             _phantom: PhantomData,
         }
     }
 }
-impl<B, C> ChainInfoProvider for SubstrateChainInfoProvider<B, C>
+impl<H, HB> ChainInfoProvider for SubstrateChainInfoProvider<H, HB>
 where
-    B: BlockT<Hash = BlockHash>,
-    B::Header: HeaderT<Number = BlockNumber>,
-    C: HeaderBackend<B>,
+    H: Header,
+    HB: HeaderBackend<H>,
 {
     fn is_block_imported(&mut self, block: &BlockId) -> bool {
-        let maybe_header = self
-            .client
-            .header(block.hash())
-            .expect("client must answer a query");
-        if let Some(header) = maybe_header {
-            // If the block number is incorrect, we treat as not imported.
-            return *header.number() == block.number();
+        match self.client.header(block) {
+            Ok(maybe_header) => maybe_header.is_some(),
+            Err(e) => {
+                log::debug!("Error while fetching header in ChainInfoProvider: {:?}", e);
+                false
+            }
         }
-        false
     }
 
-    fn get_finalized_at(&mut self, num: BlockNumber) -> Result<BlockId, ()> {
-        if self.client.info().finalized_number < num {
-            return Err(());
-        }
-
-        let block_hash = match self.client.hash(num).ok().flatten() {
-            None => {
-                error!(target: "chain-info", "Could not get hash for block #{:?}", num);
-                return Err(());
-            }
-            Some(h) => h,
-        };
-
-        if let Some(header) = self.client.header(block_hash).expect("client must respond") {
-            Ok((header.hash(), num).into())
-        } else {
-            Err(())
+    fn get_finalized_at(&mut self, number: BlockNumber) -> Result<BlockId, ()> {
+        match self.client.header_of_finalized_at(number) {
+            Ok(Some(header)) => Ok(header.id()),
+            _ => Err(()),
         }
     }
 
     fn get_parent_hash(&mut self, block: &BlockId) -> Result<BlockHash, ()> {
-        if let Some(header) = self
-            .client
-            .header(block.hash())
-            .expect("client must respond")
-        {
-            Ok(*header.parent_hash())
-        } else {
-            Err(())
+        match self.client.header(block) {
+            Ok(Some(header)) => Ok(header.parent_id().ok_or(())?.hash()),
+            Ok(None) => {
+                log::info!("Block not found while getting parent hash in ChainInfoProvider");
```

Comment:
Also missing `target: ` here and below.

Reply:
Oh, and import these rather than fully qualifying them plx.

Reply:
Target `aleph-data-store` is the best match I guess